client/gutenberg/editor/browser-url/index.js

```diff
@@ -0,0 +1,63 @@
+/**
+ * External dependencies
+ */
+import { Component } from '@wordpress/element';
+import { withSelect } from '@wordpress/data';
+
+/**
+ * After making changes to a new post, this component will update the url to append the post id:
+ *
+ * For example:
+ * https://wordpress.com/gutenberg/post/mysiteslug
+ *
+ * https://wordpress.com/gutenberg/post/mysiteslug/1234
+ */
+export class BrowserURL extends Component {
+	constructor() {
+		super( ...arguments );
+
+		this.state = {
+			historyId: null,
+		};
+	}
+
+	componentDidUpdate( prevProps ) {
+		const { postId, postStatus } = this.props;
+		const { historyId } = this.state;
+
+		if ( ( postId !== prevProps.postId || postId !== historyId ) && postStatus !== 'auto-draft' ) {
+			this.setBrowserURL( postId );
+		}
+	}
+
+	/**
+	 * Replaces the browser URL with a post editor link for the given post ID
+	 *
+	 * @param {number} postId Post ID for which to generate post editor URL.
+	 */
+	setBrowserURL( postId ) {
+		window.history.replaceState(
+			{ id: postId },
+			null,
+			`${ window.location.href }/${ postId }`
+		);
+
+		this.setState( () => ( {
+			historyId: postId,
+		} ) );
```

Comment:
we may want to rearrange this here so that the state updates first. I don't know if we could envision a re-render on `replaceState` or not but if that's possible but there's also an almost-circular thing going on here

what if we flipped it so `setState` were in `componentDidUpdate` and the `setBrowserURL` method _only_ set the browser URL?

```js
if ( … ) {
  this.setState( { historyId: postId }, () => this.setBrowserURL( postId ) );
}
```

also was the function syntax in `setState` habit or leftover? since we're assigning a raw value there I don't think there's any difference in the runtime

Reply:
I basically deleted code 🔪 from the base Gutenberg component.

The guts of the update cycle are from the `edit-post` BrowserURL component, but I suppose I could write something smaller from scratch :)

Last 30 minutes of the day trying to sneak in a PR

Reply:
Oh hmm, we have `react/no-did-update-set-state` as an eslint rule. Let me 🤔 a bit more on this.